### PR TITLE
File manager

### PR DESCRIPTION
**Pull request by david-yz-liu**
Work of @jiangha4, replacing the student file manager table with a React table.

### CODE REVIEW

File: app/assets/javascripts/file_manager.js [hunk 1]
```diff
@@ -21,37 +14,18 @@ function injectFileInput() {
...
       }
-      fileCount++;
     }
     
     return false;
   });
+  */
 
+jQuery(document).ready(function (){
+	window.modal_addnew = new ModalMarkus('#addnew_dialog');
```
Comment (arkon) [inline app/assets/javascripts/file_manager.js]: Indentation

File: app/controllers/submissions_controller.rb [hunk 2]
```diff
@@ -146,23 +148,16 @@ def populate_file_manager
...
-
+     files_array = @files.each do |file_name, file|
+       f = Hash.new
+       f[:id] = file.object_id
+       f[:file_name] = file_name
+       f[:last_modified_date] = file.last_modified_date.strftime('%d %B, %l:%M%p')
+       f[:revision_by] = file.user_id
+       f[:last_modified_revision] = file.last_modified_revision
+     end
+      # Coverts the hash to an array
```
Comment (arkon) [inline app/controllers/submissions_controller.rb]: Typo: "Converts"

File: app/views/submissions/_react_file_manager_boot_table.js.jsx.erb [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+        files: [],
+        selected_files: [],
+        data_files: [],
+      };
+    },
+    componentWillMount: function() {
+      this.refresh();
+    },
+    updateSelectedFiles: function(files) {
+    	this.setState({selected_files: files})
```
Comment (arkon) [inline app/views/submissions/_react_file_manager_boot_table.js.jsx.erb]: Missing semicolon

File: app/views/submissions/_react_file_manager_boot_table.js.jsx.erb [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+      });
+    },
+    onError: function(error) {
+      this.setState({error: error});
+    },
+    render: function() {
+      var files_data = this.state.files.map(function(file, i) {
+        var f = {};
+
+        f['id'] = i;
```
Comment (arkon) [inline app/views/submissions/_react_file_manager_boot_table.js.jsx.erb]: Dot notation is usually preferred. (E.g. `f.id = i;`)

File: app/views/submissions/_react_file_manager_boot_table.js.jsx.erb [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+        f['file_name'] = file[1].name;
+        f['last_modified_date'] = file[1].last_modified_date;
+        f['revision_by'] = file[1].user_id;
+        f['last_modified_revision'] = file[1].last_modified_revision;
+        return f;
+      }.bind(this));
+	return (
+	<div>	
+		<FileManagerActionBox
```
Comment (arkon) [inline app/views/submissions/_react_file_manager_boot_table.js.jsx.erb]: Bunch of indentation in this block

File: app/views/submissions/_react_file_manager_boot_table.js.jsx.erb [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+  });
+
+  var FileManagerActionBox = React.createClass({
+    propTypes: {
+      selected_files: React.PropTypes.array,
+      data_file: React.PropTypes.array
+    },
+    accessById: function(id) {
+      var data_file = this.props.data_file;
+      var len = data_file.length;
```
Comment (arkon) [inline app/views/submissions/_react_file_manager_boot_table.js.jsx.erb]: Does this really need to be its own variable?

File: app/views/submissions/_react_file_manager_boot_table.js.jsx.erb [hunk 7]
```diff
@@ -0,0 +1,216 @@
...
+      var data;
+      for (var i = 0; i < len; i++) {
+        if (data_file[i].id === id) {
+          data = data_file[i];
```
Comment (arkon) [inline app/views/submissions/_react_file_manager_boot_table.js.jsx.erb]: Can we just return at this point?

File: app/views/submissions/_react_file_manager_boot_table.js.jsx.erb [hunk 8]
```diff
@@ -0,0 +1,216 @@
...
+      }.bind(this));
+      return revs;
+    },
+    deleteFiles: function(e) {
+      e.preventDefault();
+      var delete_file = this.accessThroughArray();
+      var file_revisions = this.getFileRevisions();
+      jQuery.ajax({
+        url: 'update_files',
+	method: 'POST',
```
Comment (arkon) [inline app/views/submissions/_react_file_manager_boot_table.js.jsx.erb]: Indentation

File: app/views/submissions/_react_file_manager_boot_table.js.jsx.erb [hunk 9]
```diff
@@ -0,0 +1,216 @@
...
+    openModal: function(e) {
+      e.preventDefault();
+      <% if (@grouping.is_valid? && !@grouping.repository_external_commits_only?) %>
+        modal_addnew.open();
+        set_onbeforeunload(true);
+      <% end %>
+    },
+    render: function() {
+      return (
+	<div className = 'react_buttons'>
```
Comment (arkon) [inline app/views/submissions/_react_file_manager_boot_table.js.jsx.erb]: Remove spaces around `=`.

File: app/views/submissions/_react_file_manager_boot_table.js.jsx.erb [hunk 10]
```diff
@@ -0,0 +1,216 @@
...
+        </div>
+      );
+    }
+  });
+
+
+  React.renderComponent(<FileTable/>, document.getElementById('file_table'));
+</script>
+
+<script type = "text/javascript">
```
Comment (arkon) [inline app/views/submissions/_react_file_manager_boot_table.js.jsx.erb]: Remove spaces around `=`.

File: app/views/submissions/_react_file_manager_boot_table.js.jsx.erb [hunk 11]
```diff
@@ -0,0 +1,216 @@
...
+//<![CDATA[
```
Comment (arkon) [inline app/views/submissions/_react_file_manager_boot_table.js.jsx.erb]: Entire block should be indented.

File: app/views/submissions/_react_file_manager_boot_table.js.jsx.erb [hunk 12]
```diff
@@ -0,0 +1,216 @@
...
+
+  /*
+ * Do the same as what would be done server side in terms
+ * of filename sanitation. Warn the user what would happen
+ * if submit is pressed.
+ * See also: public/javascripts/file_manager.js
+ */
+function sanitized_filename_check(file_input) {
+  file_name = file_input.value.replace(/^.*[\\\/]/, '');
```
Comment (arkon) [inline app/views/submissions/_react_file_manager_boot_table.js.jsx.erb]: Need `var` declarations (below too).

File: app/views/submissions/file_manager.html.erb [hunk 13]
```diff
@@ -126,34 +126,29 @@
     </div>
   <% end %>
 
+  <% unless !@grouping.is_valid? || @grouping.repository_external_commits_only? %>
+    <%= form_tag update_files_assignment_submissions_path(@assignment.id),
+                 { multipart: true, id: 'submit_form' } do %>
+      <div id='file_table'/>
```
Comment (arkon) [inline app/views/submissions/file_manager.html.erb]: Use `<div id='file_table'></div>`.

File: app/views/submissions/file_manager.html.erb [hunk 14]
```diff
@@ -126,34 +126,29 @@
...
-                   class: 'button',
-                   onclick: "set_onbeforeunload(false); return confirm('#{I18n.t(:discard_changes_message)}');"
-                   ) unless @grouping.repository_external_commits_only? %>
-      <% end %>
-    </div>
+               { multipart: true, id: 'file_form' } do %>
+    <input type='file' name='new_files[]'>
+    <br>
+    <br>
+    <% unless !@grouping.is_valid? || @grouping.repository_external_commits_only? %>
```
Comment (arkon) [inline app/views/submissions/file_manager.html.erb]: Should this whole dialog be hidden if this is the case?

**Comment by david-yz-liu** [comment]
@arkon If you have time to make these changes, that would be great!

Also, you're particularly well-suited for the task of changing the new file modal to enable multiple file uploads at once. (Please? :D)

**Comment by arkon** [comment]
@david-yz-liu Sure. Should I submit a PR to the file-manager branch?

**Comment by david-yz-liu** [comment]
@arkon Yes, please! I'm currently trying to fix the failing tests, so this isn't going to be merged right away.

**Comment by arkon** [comment]
@david-yz-liu Do file uploads/deletions actually work for you?

**Comment by david-yz-liu** [comment]
@arkon Hmm they should be, unless I messed something up. I'll check it out.

**Comment by arkon** [comment]
@david-yz-liu Nevermind, it mysteriously works now...

**Comment by david-yz-liu** [comment]
@arkon I just tried adding, replacing, and deleting a file. Worked fine for me.

**Comment by arkon** [comment]
@david-yz-liu It was a controller path issue or something, but adding/deleting seems to work fine now. I always seem to get this when trying to replace something though:

image

**Comment by david-yz-liu** [comment]
Ah. I don't think the flash error is reset properly, so I don't think it's an actual problem with the request or controller. Also, I *did* change the logic for replacing files in the latest commit, but I will double-check to make sure this error doesn't happen again.

**Pull request merged**